### PR TITLE
Add timeout to db/storage health checks, increase reconcile concurrency 

### LINKED ISSUE #280: [Bug]: When performing DB health check takes a while, delaying creation of status update in DSPA CR
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Deploy type

Standalone DSPO (without ODH)

### Version

v1.2.0

### Environment

* SDK Version: v1.5.x
* OCP Pipelines Version: 1.11
* OCP Version: 4.13.2

### Current Behavior

When creating a DSPA I noticed the following logs: 

```
2023-08-23T16:51:01Z INFO Performing Database Health Check {"namespace": "my-ds-project", "dspa_name": "pipelines-definition"}
2023-08-23T16:51:02Z DEBUG Reconcile event triggered by [Pod: mariadb-pipelines-definition-6d848d8d75-b9cjt] {"namespace": "my-ds-project"}
2023-08-23T16:51:02Z DEBUG Reconcile event triggered by [Pod: mariadb-pipelines-definition-6d848d8d75-b9cjt] {"namespace": "my-ds-project"}
2023-08-23T16:51:02Z DEBUG Reconcile event triggered by [Pod: mariadb-pipelines-definition-6d848d8d75-b9cjt] {"namespace": "my-ds-project"}
2023-08-23T16:51:02Z DEBUG Reconcile event triggered by [Pod: mariadb-pipelines-definition-6d848d8d75-b9cjt] {"namespace": "my-ds-project"}
2023-08-23T16:51:12Z DEBUG Reconcile event triggered by [Pod: mariadb-pipelines-definition-6d848d8d75-b9cjt] {"namespace": "my-ds-project"}
2023-08-23T16:51:12Z DEBUG Reconcile event triggered by [Pod: mariadb-pipelines-definition-6d848d8d75-b9cjt] {"namespace": "my-ds-project"}
2023-08-23T16:51:13Z DEBUG Reconcile event triggered by [Pod: mariadb-pipelines-definition-6d848d8d75-b9cjt] {"namespace": "my-ds-project"}
2023-08-23T16:51:13Z DEBUG Reconcile event triggered by [Pod: mariadb-pipelines-definition-6d848d8d75-b9cjt] {"namespace": "my-ds-project"}
2023-08-23T16:51:21Z DEBUG Reconcile event triggered by [Pod: mariadb-pipelines-definition-6d848d8d75-b9cjt] {"namespace": "my-ds-project"}
2023-08-23T16:51:21Z DEBUG Reconcile event triggered by [Pod: mariadb-pipelines-definition-6d848d8d75-b9cjt] {"namespace": "my-ds-project"}
2023-08-23T16:51:33Z INFO Database Health Check Successful {"namespace": "my-ds-project", "dspa_name": "pipelines-definition"}
```

Notice the timestamp between: 
```
2023-08-23T16:51:01Z INFO Performing Database Health Check {"namespace": "my-ds-project", "dspa_name": "pipelines-definition"}
2023-08-23T16:51:33Z INFO Database Health Check Successful {"namespace": "my-ds-project", "dspa_name": "pipelines-definition"}
```

Is ~30 seconds, in this time we are taking to check whether the DB is online, but this is blocking reconcile flow and the DSPA status will be blank within this time frame. We should shorten the time out if we can for this so we do not block reconcile flow (and thus status update) for this long period of a time, if it fails in ~3 seconds, we report "Could not connect to DB" and move on, and then try again either view requeue or maridb pod changes (we already requeue automatically when mariadb pod state changes).

### Expected Behavior

DSPA Status will update faster during initial deploy.

### Steps To Reproduce

1. Deploy DSPO and DSPA
2. Watch DSPA status and notice that status won't update until mariadb pod is running 

### Workaround (if any)

none

### Anything else

_No response_

### PR DESCRIPTION
**Pull request by HumairAK**
## The issue resolved by this Pull Request:
Resolves #280
May also address #260 -- not confirmed.

## Description of your changes:
Add 1second time out to db/storage healthcheck.
Also add concurrent reconciles so when deploying multiple DSPAs they do not get backlogged (as much) in watcher queue.

## Testing instructions
1. Deploy DSPO from this pr
2. Deploy 3 dspa's simultaneously 
3. Ensure that the status field is updated in ~1seconds for all 3 (confirm 1s timeout on db/storage health check) 
4. Also ensure that pods start spinning up at the same time for all 3 (confirm concurrent reconciles)
5. Regression checks:
  * Ensure when db/s3 info is invalid the status is reported as such in the dspa CR status field 
  * Ensure valid external db/s3 configurations result in full DSPA deployment (run a flip coin to sanity check)
  * Ensure valid external s3 + default mariadb results in full dSPA deployment (run a flip coin to sanity check)

## Checklist
- [x] The commits are squashed in a cohesive manner and have meaningful messages.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

**Comment by dsp-developers** [comment]
A new image has been built to help with testing out this PR: `quay.io/opendatahub/data-science-pipelines-operator:pr-324`
An OCP cluster where you are logged in as cluster admin is required.

To use this image run the following:

```bash
cd $(mktemp -d)
git clone git@github.com:opendatahub-io/data-science-pipelines-operator.git
cd data-science-pipelines-operator/
git fetch origin pull/324/head
git checkout -b pullrequest 26307d4ffc4905e52c67151ba7d834af2d7662a9
make deploy IMG="quay.io/opendatahub/data-science-pipelines-operator:pr-324"
```

More instructions [here](https://github.com/opendatahub-io/data-science-pipelines-operator#deploy-dsp-instance) on how to deploy and test a Data Science Pipelines Application.

File: controllers/database.go [hunk 1]
```diff
@@ -38,6 +39,10 @@ var dbTemplates = []string{
 
 // extract to var for mocking in testing
 var ConnectAndQueryDatabase = func(host, port, username, password, dbname string) bool {
+	// Create a context with a timeout of 1 second
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
```
Comment (gmfrasca) [inline controllers/database.go:43]: is a 1 second timeout too aggressive?  Perhaps this should be configurable?

File: controllers/storage.go [hunk 2]
```diff
@@ -78,6 +79,9 @@ var ConnectAndQueryObjStore = func(ctx context.Context, log logr.Logger, endpoin
 		return false
 	}
 
+	ctx, cancel := context.WithTimeout(ctx, time.Second)
```
Comment (gmfrasca) [inline controllers/storage.go:82]: similar to above (maybe less so since this is s3), is a 1 second timeout too aggressive and/or should be configurable?

**Comment by dsp-developers** [comment]
Change to PR detected. A new PR build was completed.
  A new image has been built to help with testing out this PR: `quay.io/opendatahub/data-science-pipelines-operator:pr-324`

**Comment by HumairAK** [comment]
updated time out to 15s for both, example for mariadb pod scaled down (bottom) shows dspa status field (top) taking ~15 sec to update: 



issue-280

File: controllers/storage.go (hunk 2 above)
Comment (HumairAK) [reply controllers/storage.go]: updated

File: controllers/database.go (hunk 1 above)
Comment (HumairAK) [reply controllers/database.go]: updated

File: controllers/config/defaults.go [hunk 3]
```diff
@@ -103,6 +104,9 @@ var requiredFields = []string{
 	OAuthProxyImagePath,
 }
 
+// Default DB/Object storage healthcheck timeout
+const DefaultConnectionTimeout = time.Second * 15
```
[the remark this answers is not in the document]
Comment (HumairAK) [reply controllers/config/defaults.go]: resolved

File: controllers/dspipeline_controller.go [hunk 4]
```diff
@@ -573,6 +574,9 @@ func (r *DSPAReconciler) SetupWithManager(mgr ctrl.Manager) error {
 				return reconcileRequests
 			})).
 		// TODO: Add watcher for ui cluster rbac since it has no owner
+		WithOptions(controller.Options{
+			MaxConcurrentReconciles: 10,
```
Comment (gmfrasca) [inline controllers/dspipeline_controller.go:578]: should this also be a variable? (no magic numbers!)

**Comment by dsp-developers** [comment]
Change to PR detected. A new PR build was completed.
  A new image has been built to help with testing out this PR: `quay.io/opendatahub/data-science-pipelines-operator:pr-324`

File: controllers/dspipeline_controller.go (hunk 4 above)
Comment (HumairAK) [reply controllers/dspipeline_controller.go]: fixed

**Comment by dsp-developers** [comment]
Change to PR detected. A new PR build was completed.
  A new image has been built to help with testing out this PR: `quay.io/opendatahub/data-science-pipelines-operator:pr-324`

**Review by gmfrasca — COMMENTED** [review]
/lgtm

**Pull request merged**